penrose/penrose
# fix: resolve state-related bugs in the editor

# Description

Resolves the non-flaky bugs in #1683. 

(fyi: draft state = user has unsaved changes that are not tracked in a workspace. currently, this is visually indicated by the presence of a "save" button on top left side of the TopBar)
Many changes related to importing a Penrose SVG to the editor:
* added confirmation if you are trying to import a file while in draft state
* changed behavior so that the the title/top header matches the imported file name instead (previously, the title would not update at all when the file was imported)
  * b/c of above change, if you share imported trio the new title will be included
* importing a file now puts user in draft state. This fixes a bug where if you imported a file while viewing a workspace, the imported files would overwrite the workspace's changes.
* fixed bug where if you imported a trio, all the ".domain/.substance/.style" tabs would change to  "SVG import"

Other fixes:
* fixed situation where creating a gist from a draft would not save the changes locally before you're automatically redirected to the gist. Changed behavior so that the draft will be saved to a new workspace before being redirected.
* added a beforeunload event that confirms that you have no unsaved changes before navigating away from the page, addressing situations where:
  * there's no confirmation that you're abandoning work if you navigate to someone else's gist while in draft mode
  * there's no confirmation that you're abandoning work if you navigate to new webpage while in draft mode

# Implementation strategy and design decisions
I decided not to handle the flaky bugs in this PR because I believe they will require more involved changes to the recoil state, refer to issue #1695 for more information. 

# Examples with steps to reproduce them
How to get into draft state:
1. go to Examples -> make a change to any example diagram -> the blue "save" button should appear next to the diagram title

Testing changes that involve discarding a draft:
1. from draft state -> go to "upload" -> upload any Penrose SVG -> should receive a confirmation toast warning about unsaved changes
2. from draft state -> click "share" (make sure you're signed into github so you can create a gist) -> should redirect to new gist page -> press "back" button in your browser -> click "saved" tab -> draft should be listed in your workspaces
3. from draft state -> type a new website URL into your browser (i.e. google.com) -> should receive confirmation toast warning about unsaved changes

Testing changes involving importing:
1. click "upload" -> import any Penrose SVG" -> notice that:
  * "save" button should appear next to diagram title indicating that you're in  draft state
  * the diagram title should match the filename of your imported SVG
2. click "saved" -> click any of your existing workspaces -> click "upload" -> import any Penrose SVG -> click "saved" again -> notice that the workspace you had previously selected remains unchanged
 
# Checklist

- [ ] I have reviewed any generated registry diagram changes

# Open questions

Questions that require more discussion or to be addressed in future development:
